File: 2020/2020W31/PreppindataW31.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/Medals PD WOW.xlsx")
medal = pd.read_excel(xlsx,'Medallists')   
code_map = {cl[0]:cl[1] for cl in pd.read_excel(xlsx,'Country Codes').to_dict('split')['data']}
country_map = {cl[1]:cl[0] for cl in pd.read_excel(xlsx,'Country Codes').to_dict('split')['data']}

medal['Code'] = medal.apply(lambda x: code_map.get(x['Country'],x['Country Code']), axis=1)
medal['Country'] = medal.apply(lambda x: country_map.get(x['Code'],x['Country']) if pd.isna(x['Country']) else x['Country'], axis=1)
medal = medal.dropna(subset=['Country','Code']).drop('Country Code', axis=1)
medal['Event'] = medal['Event'].apply(lambda x: re.sub('(?<!kilo)(metre(s*))','m',x))
medal['Event'] = medal['Event'].apply(lambda x: re.sub('(kilometre(s*))','km',x))
medal['Sport'] = medal['Sport'].apply(lambda x: re.sub('^Canoe.*','Canoeing',x))
medal['Sport'] = medal['Sport'].apply(lambda x: re.sub('^Swimming$','Aquatics',x))  
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Beach volley.*','Beach Volleyball',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Wrestling.*','Wrestling',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Rhythmic.*','Rhythmic',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Artistic.*','Artistic',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Mountain (B|b)ik.*','Mountain Bike',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('Modern (P|p)en.*','Modern Pentath.',x))
medal['Discipline'] = medal['Discipline'] .apply(lambda x: re.sub('(.*) cycling','Cycling \\1',x))

finalA = medal[['Country', 'Code', 'Sport', 'Medal', 'Event', 'Athlete', 'Year', 'Event_Gender', 'Discipline']].copy()

finalB = medal.pivot_table(index = ['Country','Year'],
                           columns= 'Medal',
                           values= 'Athlete',
                           aggfunc={'Athlete':'count'}).reset_index()
finalB = finalB[['Country', 'Year', 'Gold', 'Silver', 'Bronze']]

finalC = pd.read_excel(xlsx,'Hosts').astype({'Start Date': str, 'End Date': str})  
finalC['Start Date'] =  finalC['Start Date'].apply(lambda x: pd.to_datetime(x))
finalC['Start Date'] =  finalC['Start Date'].dt.date
finalC['End Date'] =  finalC['End Date'].apply(lambda x: pd.to_datetime(x))
finalC['End Date'] =  finalC['End Date'].dt.date
finalC['Year'] = finalC['Start Date'].apply(lambda x: x.year)
temp = finalC['Host'].str.split(',', n = 1, expand = True)
finalC['Host City'] = temp[0].str.strip()
finalC['Host Country'] = temp[1].str.strip()
finalC = finalC[['Year', 'Host Country', 'Host City', 'Start Date', 'End Date', 'Games', 'Nations', 'Sports', 'Events']].copy()

File: 2020/2020W31/preppindataw31.R
library(dplyr)
library(readxl)
library(tidyr)
library(stringr)

xlsx <- "F:/Data/Medals PD WOW.xlsx"

country_code_map <- read_excel(xlsx, sheet = 'Country Codes') 

finalA <- read_excel(xlsx, sheet = 'Medallists', col_types = c("text", 
                                                               "text", "text", "text", "text", "text", 
                                                               "text", "text", "text")) %>%
  merge(country_code_map, by = 'Country', all.x = TRUE) %>%
  mutate('Code' = if_else(is.na(`Code`),`Country Code`,`Code`)) %>%
  rename('CountryDrop' = 'Country') %>%
  merge(country_code_map, by = 'Code') %>%
  mutate('Event' = str_replace_all(`Event`,'(?<!kilo)(metre(s*))','m'),
         'Event' = str_replace_all(`Event`,'(kilometre(s*))','km'),
         'Sport' = str_replace_all(`Sport`,'^Canoe.*','Canoeing'),
         'Sport' = str_replace_all(`Sport`,'^Swimming$','Aquatics'),
         'Discipline' = str_replace_all(`Discipline`,'Beach volley.*','Beach Volleyball'),
         'Discipline' = str_replace_all(`Discipline`,'Wrestling.*','Wrestling'),
         'Discipline' = str_replace_all(`Discipline`,'Rhythmic.*','Rhythmic'),
         'Discipline' = str_replace_all(`Discipline`,'Artistic.*','Artistic'),
         'Discipline' = str_replace_all(`Discipline`,'Mountain (B|b)ik.*','Mountain Bike'),
         'Discipline' = str_replace_all(`Discipline`,'Modern (P|p)en.*','Modern Pentath.'),
         'Discipline' = str_replace_all(`Discipline`,'(.*) cycling','Cycling \\1')) %>%
  select(c('Country', 'Code', 'Sport', 'Medal', 'Event', 'Athlete', 'Year', 'Event_Gender', 'Discipline'))

finalB <- finalA %>% 
  group_by(`Country`, `Year`, `Medal`) %>%
  summarise('Value' = n()) %>%
  pivot_wider(., names_from = `Medal`, values_from = `Value`, values_fn = list(Value=sum)) %>%
  select(c('Country', 'Year', 'Gold', 'Silver', 'Bronze'))

finalC <- read_excel(xlsx, sheet = 'Hosts',col_types = c("text", 
                                                         "text", "text", "text", "numeric", 
                                                         "numeric", "numeric")) %>%
  separate(., `Host`, c('Host City', 'Host Country'), sep = ',\\s') %>%
  mutate('Start Date' = if_else(str_detect(`Start Date`,'/'),
                                as.Date(`Start Date`,format='%m/%d/%Y'),
                                as.Date(as.numeric(`Start Date`), origin = '1899-12-30')),
         'End Date' = if_else(str_detect(`End Date`,'/'),
                                as.Date(`End Date`,format='%m/%d/%Y'),
                                as.Date(as.numeric(`End Date`), origin = '1899-12-30')),
         'Year' = as.integer(strftime(`Start Date`,format = '%Y'))) %>%
  select(c('Year', 'Host Country', 'Host City', 'Start Date', 'End Date', 'Games', 'Nations', 'Sports', 'Events'))

View(finalA)
View(finalB)
View(finalC)